rechunker/algorithm.py

```diff
@@ -91,38 +93,107 @@ def consolidate_chunks(
     return tuple(new_chunks)
 
 
-def rechunking_plan(
+def _calculate_shared_chunks(
+    read_chunks: Sequence[int], write_chunks: Sequence[int]
+) -> Tuple[int, ...]:
+    # Intermediate chunks are the smallest possible chunks which fit
+    # into both read_chunks and write_chunks.
+    # Example:
+    #   read_chunks:            (20, 5)
+    #   target_chunks:          (4, 25)
+    #   intermediate_chunks:    (4, 5)
+    # We don't need to check their memory usage: they are guaranteed to be smaller
+    # than both read and write chunks.
+    return tuple(
+        min(c_read, c_target) for c_read, c_target in zip(read_chunks, write_chunks)
+    )
+
+
+def calculate_stage_chunks(
+    read_chunks: Tuple[int, ...],
+    write_chunks: Tuple[int, ...],
+    stage_count: int = 1,
+    epsilon: float = 1e-8,
+) -> List[Tuple[int, ...]]:
+    """
+    Calculate chunks after each stage of a multi-stage rechunking.
+
+    Each stage consists of "split" step followed by a "consolidate" step.
+
+    The strategy used here is to progressively enlarge or shrink chunks along
+    each dimension by the same multiple in each stage. This should roughly
+    minimize the total number of arrays resulting from "split" steps in a
+    multi-stage pipeline. It also keeps the total number of elements in each
+    chunk constant, up to rounding error, so memory usage should remain
+    constant.
+
+    Examples::
+
+        >>> calculate_stage_chunks((1_000_000, 1), (1, 1_000_000), stage_count=2)
+        [(1000, 1000)]
+        >>> calculate_stage_chunks((1_000_000, 1), (1, 1_000_000), stage_count=3)
+        [(10000, 100), (100, 10000)]
+        >>> calculate_stage_chunks((1_000_000, 1), (1, 1_000_000), stage_count=4)
+        [(31623, 32), (1000, 1000), (32, 31623)]
+
+    TODO: consider more sophisticated algorithms.
+    """
+    stage_chunks = []
+    for stage in range(1, stage_count):
+        power = stage / stage_count
+        # Add a small floating-point epsilon so we don't inadvertently
+        # round-down even chunk-sizes.
+        chunks = tuple(
+            floor(rc ** (1 - power) * wc**power + epsilon)
+            for rc, wc in zip(read_chunks, write_chunks)
+        )
+        stage_chunks.append(chunks)
+    return stage_chunks
+
+
+def _count_num_splits(source_chunk: int, target_chunk: int, size: int) -> int:
+    multiple = lcm(source_chunk, target_chunk)
+    splits_per_lcm = multiple // source_chunk + multiple // target_chunk - 1
+    lcm_count, remainder = divmod(size, multiple)
+    if remainder:
+        splits_in_remainder = (
+            ceil(remainder / source_chunk) + ceil(remainder / target_chunk) - 1
+        )
+    else:
+        splits_in_remainder = 0
+    return lcm_count * splits_per_lcm + splits_in_remainder
+
+
+def calculate_single_stage_io_ops(
+    shape: Sequence[int], in_chunks: Sequence[int], out_chunks: Sequence[int]
+) -> int:
+    """Estimate the number of irregular chunks required for rechunking."""
```

Comment:
I cannot reconcile the docstring with the function name. Where are "irregular chunks" being calculated?

Reply:
removed the note about irregular chunks -- that's really unrelated.